src/analysis/convex/combination.lean

```diff
@@ -7,162 +7,184 @@ import analysis.convex.basic
 import algebra.big_operators.order
 
 /-!
-# Convex combinations
+# Linear combinations
 
-This file defines convex combinations of points in a vector space.
+This file defines linear combinations of points in a semimodule and proves results about convex
+combinations.
+
+Convex combinations in a semimodule can be modelled as linear combinations with nonnegative
+coefficients summing to `1`. In an affine space, they can be modelled through affine combinations.
+
+In a vector space, both coincide but it is still an open question whether we can make the two ways
+to derive `convex_combination` definitionally equal.
 
 ## Main declarations
 
-* `finset.center_mass`: Center of mass of a finite family of points.
+* `finset.linear_combination`: Center of mass of a finite family of points.
+
+## TODO
 
-## Implementation notes
+Change `finset.linear_combination : finset ι → (ι → E) → (ι → 𝕜) → E` to
+`linear_combination : (ι → E) →ₗ[𝕜] (ι →₀ 𝕜) →ₗ[𝕜] E`. Same goes for `finset.affine_combination`.
 
-We divide by the sum of the weights in the definition of `finset.center_mass` because of the way
-mathematical arguments go: one doesn't change weights, but merely adds some. This also makes a few
-lemmas unconditional on the sum of the weights being `1`.
+Ultimately, this file should be about `convex_combination`, which will generalize both
+`affine_combination` and `linear_combination`. The latter should find a home in `linear_algebra`.
 -/
 
 open set
 open_locale big_operators classical
 
-universes u u'
-variables {R E F ι ι' : Type*}
-  [linear_ordered_field R] [add_comm_group E] [module R E] [add_comm_group F] [module R F]
-  {s : set E}
+namespace finset
 
-/-- Center of mass of a finite collection of points with prescribed weights.
-Note that we require neither `0 ≤ w i` nor `∑ w = 1`. -/
-def finset.center_mass (t : finset ι) (w : ι → R) (z : ι → E) : E :=
-(∑ i in t, w i)⁻¹ • (∑ i in t, w i • z i)
+/-- Linear combination of a finite collection of points with prescribed weights. -/
+def linear_combination {𝕜 E ι : Type*} [ordered_semiring 𝕜] [add_comm_monoid E] [module 𝕜 E]
```

Comment:
I agree that `finset.center_mass` should probably be removed/replaced but I would have expected the replacement to be `finset.affine_combination`, rather than what is proposed here. I'm not even sure I prefer `s.linear_combination p w` over `∑ i in s, w i • p i`.

Do you think you could achieve your aims by using `finset.affine_combination` instead?

Reply:
Same problem as everywhere else, we need `convex_space` (and here `convex_combination`) to go to the bottom of the convexity generalization. I'm generalizing along the semimodule path and you want me to generalize along the affine path. My choice here is that the proofs are basically unchanged, while that won't be the case using `finset.affine_combination`.

Reply:
That's a good point about not wanting to change proofs but I think using [this lemma](https://github.com/leanprover-community/mathlib/pull/9103/files#diff-ece920d3095b577efda5ca995226b64d018390dfbefdb264722220712d38682bR254) the changes should be tiny if we went with `finset.affine_combination`?

Failing that, I think I prefer a raw `∑ i in s, w i • p i` over the proposed `finset.linear_combination`.

Reply:
You could even spell it `s.sum (w • p)`

Reply:
That was my idea as a second stage refactor. If you agree, I can refactor directly like that, using `finsupp.total`.

Reply:
This conversation can now be marked as resolved, right?

Reply:
Yes!